packages/core/src/external/cda/process-attachments.ts

```diff
@@ -0,0 +1,232 @@
+import { Config } from "../../util/config";
+import { out } from "../../util/log";
+import { sizeInBytes } from "../../util/string";
+import { uuidv4 } from "../../util/uuid-v7";
+import { S3Utils, UploadParams } from "../aws/s3";
+import { cqExtension } from "../carequality/extension";
+import { cwExtension } from "../commonwell/extension";
+import { makeFhirApi } from "../fhir/api/api-factory";
+import { convertCollectionBundleToTransactionBundle } from "../fhir/bundle/convert-to-transaction-bundle";
+import { buildDocIdFhirExtension } from "../fhir/shared/extensions/doc-id-extension";
+
+const s3Utils = new S3Utils(Config.getAWSRegion());
+
+export async function processAttachments({
+  b64Attachments,
+  cxId,
+  patientId,
+  filePath,
+  medicalDataSource,
+  s3BucketName,
+  fhirUrl,
+}: {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  b64Attachments: any[];
+  cxId: string;
+  patientId: string;
+  filePath: string;
+  medicalDataSource?: string | undefined;
+  s3BucketName: string;
+  fhirUrl?: string;
+}) {
+  const { log } = out(`processAttachments - cxId ${cxId}, patientId ${patientId}`);
+
+  const extensions = [buildDocIdFhirExtension(filePath), getSourceExtension(medicalDataSource)]
+    .flat()
+    .filter(Boolean) as Extension[];
+
+  const docRefs: DocumentReference[] = [];
+  const uploadDetails: UploadParams[] = [];
+
+  b64Attachments.map(async att => {
+    const act = att.act as ConcernActEntryAct;
+
+    const fileDetails = getDetails(act.text);
+    const docRef = buildDocumentReference(patientId, extensions, act);
+    if (!docRef.id) throw new Error("Missing ID in DocRef");
+
+    const fileKey = createAttachmentUploadFileName({
+      filePath,
+      attachmentId: docRef.id,
+      mimeType: fileDetails.mimeType,
+    });
+
+    const fileUrl = s3Utils.buildFileUrl(s3BucketName, fileKey);
+
+    if (!fileDetails.fileB64Contents) return;
+    const attachment: Attachment = {
+      ...(fileDetails.mimeType && { contentType: fileDetails.mimeType }),
+      url: fileUrl,
+      size: sizeInBytes(fileDetails.fileB64Contents),
+      title: fileKey,
+    };
+
+    if (docRef.date) attachment.creation = docRef.date;
+    docRef.content = [{ attachment }];
+
+    docRefs.push(docRef);
+
+    const uploadParams = {
+      bucket: s3BucketName,
+      key: fileKey,
+      file: Buffer.from(fileDetails.fileB64Contents, "base64"),
+      ...(fileDetails.mimeType && { contentType: fileDetails.mimeType }),
+    };
+    uploadDetails.push(uploadParams);
+  });
+
+  log(`Extracted ${docRefs.length} attachments`);
+  const docRefBundleEntries = docRefs.map(dr => ({ resource: dr }));
+  const collectionBundle: Bundle = {
+    resourceType: "Bundle",
+    type: "collection",
+    entry: docRefBundleEntries,
+  };
+
+  const transactionBundle = convertCollectionBundleToTransactionBundle({
+    fhirBundle: collectionBundle,
+  });
+
+  if (transactionBundle.entry?.length && fhirUrl && s3BucketName.length) {
```

Comment:
I'd maybe check fhir and s3 separately... maybe we want one but not the other?

also, `s3BucketName.length` is a strange -- are we allowing an empty string to signify no bucket? We should simply allow undefined.

Reply:
true that

Reply:
I made some changes just now.. Not too happy with em myself, but I think it kinda makes sense. If bucket name is undefined, we will fill it in with an empty string, and then not do anything for s3 upload anyway, so that shouldn't affect anything. Besides, the lambda still requires that the bucket name is provided. So maybe this is fine?